contracts/cw721-base/src/contract.rs

```diff
@@ -0,0 +1,499 @@
+        name,
+        description: description.unwrap_or_default(),
+        image,
+    };
+    tokens(&mut deps.storage).save(token_id.as_bytes(), &token)?;
+
+    increment_tokens(&mut deps.storage)?;
+
+    // TODO: set logs
+    Ok(HandleResponse::default())
+}
+
+pub fn handle_transfer_nft<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    recipient: HumanAddr,
+    token_id: String,
+) -> StdResult<HandleResponse> {
+    let mut token = tokens(&mut deps.storage).load(token_id.as_bytes())?;
+    // ensure we have permissions
+    check_can_send(&deps, &env, &token)?;
+    // set owner and remove existing approvals
+    token.owner = deps.api.canonical_address(&recipient)?;
+    token.approvals = vec![];
+    tokens(&mut deps.storage).save(token_id.as_bytes(), &token)?;
+
+    // TODO: set logs
+    Ok(HandleResponse::default())
+}
+
+pub fn handle_send_nft<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    contract: HumanAddr,
+    token_id: String,
+    _msg: Option<Binary>,
+) -> StdResult<HandleResponse> {
+    let mut token = tokens(&mut deps.storage).load(token_id.as_bytes())?;
+    // ensure we have permissions
+    check_can_send(&deps, &env, &token)?;
+    // set owner and remove existing approvals
+    token.owner = deps.api.canonical_address(&contract)?;
+    token.approvals = vec![];
+    tokens(&mut deps.storage).save(token_id.as_bytes(), &token)?;
+
+    // TODO: set logs
+    // TODO: send msg to contract
+    Ok(HandleResponse::default())
+}
+
+pub fn handle_approve<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    spender: HumanAddr,
+    token_id: String,
+    expires: Option<Expiration>,
+) -> StdResult<HandleResponse> {
+    // reject expired data as invalid
+    let expires = expires.unwrap_or_default();
+    if expires.is_expired(&env.block) {
+        return Err(StdError::generic_err(
+            "Cannot set approval that is already expired",
+        ));
+    }
+
+    let mut token = tokens(&mut deps.storage).load(token_id.as_bytes())?;
+    // ensure we have permissions
+    check_can_approve(&deps, &env, &token)?;
+
+    // update the approval list (remove any for the same spender before adding)
+    let spender_raw = deps.api.canonical_address(&spender)?;
+    token.approvals = token
+        .approvals
+        .into_iter()
+        .filter(|apr| apr.spender != spender_raw)
+        .collect();
+    let approval = Approval {
+        spender: spender_raw,
+        expires,
+    };
+    token.approvals.push(approval);
+    tokens(&mut deps.storage).save(token_id.as_bytes(), &token)?;
+
+    // TODO: set logs
+    Ok(HandleResponse::default())
+}
+
+pub fn handle_revoke<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    spender: HumanAddr,
+    token_id: String,
+) -> StdResult<HandleResponse> {
+    let mut token = tokens(&mut deps.storage).load(token_id.as_bytes())?;
```

Comment:
Not sure it's worth it, but this logic is shared between `approve` and `revoke`. In fact, `approve` is just `revoke` plus adding a new spender.